crates/driver/src/domain/competition/solution/fee.rs

```diff
@@ -0,0 +1,226 @@
+                    let price_improvement_fee = self.price_improvement_fee(prices, *factor)?;
+                    let max_volume_fee = self.volume_fee(prices, *max_volume_factor)?;
+                    // take the smaller of the two
+                    protocol_fee = std::cmp::min(price_improvement_fee, max_volume_fee);
+                }
+                FeePolicy::Volume { factor } => {
+                    protocol_fee = self.volume_fee(prices, *factor)?;
+                }
+            }
+        }
+        Ok(protocol_fee)
+    }
+
+    fn price_improvement_fee(
+        &self,
+        prices: ClearingPrices,
+        factor: f64,
+    ) -> Result<eth::U256, Error> {
+        let sell_amount = self.order().sell.amount.0;
+        let buy_amount = self.order().buy.amount.0;
+        let executed = self.executed().0;
+        let surplus_fee = self
+            .dynamic_fee()
+            .map(|fee| fee.0)
+            .ok_or(Error::ProtocolFeeOnStaticOrder)?;
+        match self.order().side {
+            Side::Buy => {
+                // How much `sell_token` we need to sell to buy `executed` amount of `buy_token`
+                let executed_sell_amount = executed
+                    .checked_mul(prices.buy)
+                    .ok_or(Error::Overflow)?
+                    .checked_div(prices.sell)
+                    .ok_or(Error::DivisionByZero)?;
+                // Sell slightly more `sell_token` to capture the `surplus_fee`
+                let executed_sell_amount_with_fee = executed_sell_amount
+                    .checked_add(surplus_fee)
+                    .ok_or(Error::Overflow)?;
+                // Scale to support partially fillable orders
+                let limit_sell_amount = sell_amount
+                    .checked_mul(executed)
+                    .ok_or(Error::Overflow)?
+                    .checked_div(buy_amount)
+                    .ok_or(Error::DivisionByZero)?;
+                // Remaining surplus after fees
+                // Do not return error if `checked_sub` fails because violated limit prices will
+                // be caught by simulation
+                let surplus = limit_sell_amount
+                    .checked_sub(executed_sell_amount_with_fee)
+                    .unwrap_or(eth::U256::zero());
+                Ok(surplus
+                    .checked_mul(eth::U256::from_f64_lossy(factor * 100.))
+                    .ok_or(Error::Overflow)?
+                    / 100)
+            }
+            Side::Sell => {
+                // How much `buy_token` we get for `executed` amount of `sell_token`
+                let executed_buy_amount = executed
+                    .checked_mul(prices.sell)
+                    .ok_or(Error::Overflow)?
+                    .checked_div(prices.buy)
+                    .ok_or(Error::DivisionByZero)?;
+                let executed_sell_amount_with_fee =
+                    executed.checked_add(surplus_fee).ok_or(Error::Overflow)?;
```

Comment:
By adding this here `executed_sell_amount_with_fee` will be greater than `sell_amount` which means `limit_buy_amount` will be greater than `executed_buy_amount` which means `surplus` will always be negative.
Unless I got it twisted.

Reply:
Did a test computation and it looked okay.
I think I overlooked that even if you want to fill a `sell` order completely you would not set `executed == 100%` as a solver **AND** set a `surplus_fee` because then you'd spend more `sell_token` than what the user signed.
If the solver sets `executed = 100% - surplus_fee` and set `surplus_fee = some_amount` like they need to things work out.

Reply:
This is enforced in the Fulfillment constructor.